### PR TITLE
Add the Launcher Application tests.

### PR DESCRIPTION
**Pull request by zcervink**
Add the Launcher Application tests. Needs my new class from jbosstools-central, where I created pull request a moment ago.

Signed-off-by: Zbynek Cervinka <zcervink@redhat.com>

### CODE REVIEW

**Comment by zcervink** [comment]
Tested on Jenkins:
snimek obrazovky 2018-12-18 v 12 12 19

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+			fail("ParseException: unable to parse file at path is: " + pathToFile);
+		}
+	}
+
+	@Test
+	public void createNewLauncherApplicationProjectWithDefaultSettings() {
+		ArchetypeProject project = new org.jboss.tools.central.test.ui.reddeer.projects.LauncherApplication();
+
+		log.step("Import project start");
+		JBossCentralProjectWizardLauncherApplication dialog = new JBossCentralProjectWizardLauncherApplication(project);
```
Comment (jkopriva) [inline tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: project variable should not be needed to create JBossCentralProjectWizardLauncherApplication - you can delete Archetype project completely from this class and instead archetype project use only constant "Launcher Application" to simplify code

File: tests/org.jboss.tools.central.ui.bot.test/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -17,7 +17,8 @@ Require-Bundle: org.eclipse.ui.forms;bundle-version="3.5.101",
  org.apache.commons.lang;bundle-version="2.6.0",
  org.jboss.tools.common.reddeer;bundle-version="3.10.0",
  org.jboss.tools.maven.reddeer,
- org.jboss.ide.eclipse.as.reddeer
+ org.jboss.ide.eclipse.as.reddeer,
```
Comment (jkopriva) [inline tests/org.jboss.tools.central.ui.bot.test/META-INF/MANIFEST.MF]: duplicate dependency

Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/META-INF/MANIFEST.MF]: removed.

**Comment by zcervink** [comment]
JIRA created here:
https://issues.jboss.org/browse/JBIDE-26520

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ *
+ * @author zcervink
+ *
+ */
```
[the remark this answers is not in the document]
Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: JIRA added to this PR as a comment. Do I need to add the link to the JIRA anywhere else?

[the remark this answers is not in the document]
Comment (jkopriva) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: @zcervink Please add it in javadoc of class, and also it is good practice to add JBIDE-26520 in commit message.

File: tests/org.jboss.tools.central.ui.bot.test/META-INF/MANIFEST.MF (hunk 2 above)
Comment (jkopriva) [reply tests/org.jboss.tools.central.ui.bot.test/META-INF/MANIFEST.MF]: thx

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: Added to the Java file and to the commit message too.

[the remark this answers is not in the document]
Comment (jkopriva) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: thx!

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+@RunWith(RedDeerSuite.class)
+public class LauncherApplicationTest {
+
+	private static final String CENTRAL_LABEL = "Red Hat Central";
+	private static final String MAVEN_SETTINGS_PATH = System.getProperty("maven.config.file") == null
+			? "./target/classes/settings.xml"
+			: System.getProperty("maven.config.file");;
```
[the remark this answers is not in the document]
Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: removed and tested on localhost. Now testing on Jenkins.

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java (hunk 1 above)
Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: Deleted and tested on localhost.

Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: Tested on Jenkins, works good.

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+		for (Project p : new ProjectExplorer().getProjects()) {
+			p.delete(true);
+		}
+		new DefaultToolItem(new WorkbenchShell(), CENTRAL_LABEL).click();
+		// activate central editor
+		new DefaultEditor(CENTRAL_LABEL);
+	}
+
+	// https://www.mkyong.com/java/json-simple-example-read-and-write-json
+	private static void loadBlacklistErrorsFile(String blacklisterrorsFile) {
```
[the remark this answers is not in the document]
Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: Solved by changing private modifier in ArchetypesTest.java to public and successfully tested on localhost. Now testing on Jenkins.

[the remark this answers is not in the document]
Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: Tested on Jenkins, works fine.

**Comment by zcervink** [comment]
@jkopriva all required changes were made. PR on jbosstools-central needs to be merged first. Tested on Jenkins: 

jenkins

File: tests/org.jboss.tools.central.ui.bot.test/resources/blacklist-launcher-application-errors-regexes.json [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{
+	"LauncherApplicationTest": [
+		".*The element type \"p\" must be terminated by the matching end-tag \"</p>\".*",
+		".*java\.lang\.Object cannot be resolved.*",
+		".*Cannot find the class file for java\.lang\.Object.*"
```
Comment (jkopriva) [inline tests/org.jboss.tools.central.ui.bot.test/resources/blacklist-launcher-application-errors-regexes.json]: @zcervink please remove this java.lan.Object things - as you said, its issue on your local machine related to Java installation

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/projects/LauncherApplication.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.jboss.tools.central.reddeer.projects.ArchetypeProject;
+
+/**
+ *
+ * @author zcervink
+ *
+ */
+
+public class LauncherApplication extends ArchetypeProject {
```
Comment (jkopriva) [inline tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/projects/LauncherApplication.java]: @zcervink I have not found, where is this class used - if it is not used, remove it. Thank you

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+	private static final Logger log = Logger.getLogger(ExamplesOperator.class);
+	private static final String BLACKLIST_ERRORS_REGEXES_FILE = "resources/blacklist-launcher-application-errors-regexes.json";
+
+	@BeforeClass
+	public static void setup() {
+		WorkbenchPreferenceDialog preferenceDialog = new WorkbenchPreferenceDialog();
+		preferenceDialog.open();
+		MavenSettingsPreferencePage prefPage = new MavenSettingsPreferencePage(preferenceDialog);
+		preferenceDialog.select(prefPage);
+		preferenceDialog.ok();
```
Comment (jkopriva) [inline tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: @zcervink lines  60-64 were related to setting settings.xml in preferences, and now this code is useless, so please remove it.

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/projects/LauncherApplication.java (hunk 7 above)
Comment (zcervink) [reply tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/projects/LauncherApplication.java]: deleted.

**Comment by zcervink** [comment]
All changes were made. Tested on Jenkins and it is working:
jenkins

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+		log.step("Import project start");
+		
+		NewLauncherProjectWizard wizard = new NewLauncherProjectWizard();
+		NewLauncherProjectWizardPage wizardPage = new NewLauncherProjectWizardPage("New Launcher project");
+		
+		wizard.openWizardFromShellMenu();
+		wizardPage.setProjectName("project_1_default_settings");
+
+		wizardPage.clickTheFinishButton();
+		new WaitWhile(new JobIsRunning(), TimePeriod.getCustom(2500));
```
Comment (jkopriva) [inline tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: you can use wizardPage.finish(TimePeriod.getCustom(2500));

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+		wizardPage.setTargetRuntime("vert.x redhat");
+		wizardPage.setProjectName("project_2_customized_settings");
+		wizardPage.toggleUseDefaultLocationCheckBox(true);
+		wizardPage.setCustomProjectLocation("/Users/zcervink/zc-custom-loc");
+		wizardPage.setArtifactId("my_customized_artifact_id");
+		wizardPage.setGroupId("my.customized.group.id");
+		wizardPage.setVersion("1.2.3-SNAPSHOT");
+
+		wizardPage.clickTheFinishButton();
+		new WaitWhile(new JobIsRunning(), TimePeriod.getCustom(2500));
```
Comment (jkopriva) [inline tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: you can use wizardPage.finish(TimePeriod.getCustom(2500));

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+	public void createNewLauncherApplicationProjectWithDefaultSettings() {
+		log.step("Import project start");
+		
+		NewLauncherProjectWizard wizard = new NewLauncherProjectWizard();
+		NewLauncherProjectWizardPage wizardPage = new NewLauncherProjectWizardPage("New Launcher project");
+		
+		wizard.openWizardFromShellMenu();
+		wizardPage.setProjectName("project_1_default_settings");
+
+		wizardPage.clickTheFinishButton();
```
Comment (odockal) [inline tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: As was mentioned in your reddeer test-framework plugin, finish method should be implemented over wizard dialog object, not wizardpage object that accommodates form page object like labels and set up button. wizard dialog is suitable for control button like finish, next, etc...

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+		wizardPage.setTargetMission("rest-http");
+		wizardPage.setTargetRuntime("vert.x redhat");
+		wizardPage.setProjectName("project_2_customized_settings");
+		wizardPage.toggleUseDefaultLocationCheckBox(true);
+		wizardPage.setCustomProjectLocation("/Users/zcervink/zc-custom-loc");
+		wizardPage.setArtifactId("my_customized_artifact_id");
+		wizardPage.setGroupId("my.customized.group.id");
+		wizardPage.setVersion("1.2.3-SNAPSHOT");
+
+		wizardPage.clickTheFinishButton();
```
Comment (odockal) [inline tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: See comment above.

**Comment by zcervink** [comment]
All issues are solved now and tested on Jenkins:
test

File: tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+	@Test
+	public void testFinishButtonDisabledInLauncherApplicationProjectWizard() {
+		log.step("Import project start");
+
+		NewLauncherProjectWizard wizard = new NewLauncherProjectWizard();	
+		NewLauncherProjectWizardPage wizardPage = new NewLauncherProjectWizardPage(wizard);
+		wizard.openWizardFromShellMenu();
+
+		// PROJECT NAME not filled test
+		assertFalse(wizardPage.isFinishButtonEnabled());
```
Comment (odockal) [inline tests/org.jboss.tools.central.ui.bot.test/src/org/jboss/tools/central/test/ui/reddeer/LauncherApplicationTest.java]: Also, please move isFinishButtonEnabled from page into wizard dialog class same as impl. of finish methods. Thanks.

**Comment by zcervink** [comment]
Everything fixed and tested on Jenkins.

**Comment by jkopriva** [comment]
testPR

**Comment by jkopriva** [comment]
testPR

**Comment by jkopriva** [comment]
testPR

**Pull request merged**